src/adaptors/adaptor_tls.c

```diff
@@ -0,0 +1,630 @@
+            DEQ_INSERT_TAIL(*decrypted_buffs, decrypted_adaptor_buff);
+            qd_log(tls->log_source,
+                   QD_LOG_TRACE,
+                   "[C%" PRIu64 "] qd_tls_decrypt pn_tls_take_decrypt_output_buffers, decrypt adaptor buffer size=%zu",
+                   tls->conn_id,
+                   qd_adaptor_buffer_size(decrypted_adaptor_buff));
+        }
+
+        // You can take only a maximum of decrypt_output_buff_capacity. You will have to recheck
+        // pn_tls_need_decrypt_output_buffers to get everything back.
+        if (pn_tls_need_decrypt_output_buffers(tls->tls_session)) {
+            goto give_decrypt_output_buffers;
+        }
+        take_back_input_decrypt_buffs(tls);
+        fetch_size = get_pn_raw_buffer_fetch_size(tls);
+    }
+
+    tls->tls_has_output = pn_tls_is_encrypt_output_pending(tls->tls_session);
+    qd_log(tls->log_source,
+           QD_LOG_TRACE,
+           "[C%" PRIu64 "] qd_tls_decrypt tls->tls_has_output=%i",
+           tls->conn_id,
+           qd_tls_has_output(tls));
+    take_back_input_decrypt_buffs(tls);
+    return encrypted_bytes_in;
+}
+
+int qd_tls_encrypt(qd_tls_t *tls, qd_adaptor_buffer_t *unencrypted_buff, qd_adaptor_buffer_list_t *encrypted_buffs)
+{
+    if (tls->tls_error)
+        return QD_TLS_ERROR;
+
+    int unencrypted_bytes_out = 0;
+
+    if (unencrypted_buff) {
+        unencrypted_bytes_out += qd_adaptor_buffer_size(unencrypted_buff);
+        qd_log(tls->log_source,
+               QD_LOG_DEBUG,
+               "[C%" PRIu64 "] qd_tls_encrypt unencrypted_buff.size=%zu",
+               tls->conn_id,
+               qd_adaptor_buffer_size(unencrypted_buff));
+    } else {
+        qd_log(tls->log_source, QD_LOG_DEBUG, "[C%" PRIu64 "] qd_tls_encrypt no unencrypted buff sent", tls->conn_id);
+    }
+
+    if (pn_tls_is_secure(tls->tls_session)) {
+        qd_log(tls->log_source, QD_LOG_TRACE, "[C%" PRIu64 "] qd_tls_encrypt tls session is secure", tls->conn_id);
+
+        if (unencrypted_buff) {
+            size_t encrypt_input_buff_capacity = pn_tls_get_encrypt_input_buffer_capacity(tls->tls_session);
+            (void) encrypt_input_buff_capacity;  // prevent unused variable warning
+            assert(encrypt_input_buff_capacity > 0);
+
+            pn_raw_buffer_t pn_raw_buffer;
+            qd_adaptor_buffer_pn_raw_buffer(&pn_raw_buffer, unencrypted_buff);
+            size_t consumed = pn_tls_give_encrypt_input_buffers(tls->tls_session, &pn_raw_buffer, 1);
+            (void) consumed;  // prevent unused variable warning
+            assert(consumed == 1);
+            qd_log(tls->log_source,
+                   QD_LOG_TRACE,
+                   "[C%" PRIu64 "] qd_tls_encrypt pn_tls_give_encrypt_input_buffers success",
+                   tls->conn_id);
+
+            //
+            // Process TLS.
+            //
+            if (!process_tls(tls)) {
+                take_back_input_encrypt_buffs(tls);
```

Comment:
Shouldn't this be called on entry to make room for the new input encrypted buffs?